gcp_variant_transforms/libs/annotation/vep/vep_runner.py

```diff
@@ -0,0 +1,464 @@
+    # operation instead of just checking `done`.
+    if is_done:
+      logging.info('Operation %s is done.', operation)
+    return is_done
+
+  def run_on_all_files(self):
+    # type: () -> None
+    """Runs VEP on all input files.
+
+    The input files structure is recreated under `self._output_dir` and each
+    output file will have `_VEP_OUTPUT_SUFFIX`.
+    """
+    if self._running_operation_ids:
+      raise AssertionError('There are already {} operations running.'.format(
+          len(self._running_operation_ids)))
+    logging.info('Finding all files that match %s', self._input_pattern)
+    match_results = filesystems.FileSystems.match([
+        self._input_pattern])  # type: List[filesystem.MatchResult]
+    if not match_results:
+      raise ValueError('No files matched input_pattern: {}'.format(
+          self._input_pattern))
+    logging.info('Number of files: %d', len(match_results[0].metadata_list))
+    self._check_and_write_to_output_dir(self._output_dir)
+    pipelines_data = _PipelinesExecutionInfo(
+        match_results[0].metadata_list, self._output_dir, self._max_num_workers)
+    for vm_index, actions in enumerate(pipelines_data.single_vm_actions_list):
+      operation_name = self._call_pipelines_api(
+          actions, self._get_output_log_path(self._output_dir, vm_index))
+      logging.info('Started operation %s on VM %d processing %d input files',
+                   operation_name, vm_index, len(actions.io_map))
+      self._running_operation_ids.append(operation_name)
+
+  def _call_pipelines_api(self, single_vm_actions, output_log_path):
+    # type: (_SingleWorkerActions, str) -> str
+    api_request = self._get_api_request_fixed_parts()
+    size_gb = single_vm_actions.disk_size_bytes / (1 << 30)
+    api_request[_API_PIPELINE]['resources'][
+        'virtualMachine']['disks'][0]['sizeGb'] = (
+            size_gb + _MINIMUM_DISK_SIZE_GB)
+    for input_file, output_file in single_vm_actions.io_map.iteritems():
+      api_request[_API_PIPELINE][_API_ACTIONS].extend(
+          self._create_actions(input_file, output_file))
+    api_request[_API_PIPELINE][_API_ACTIONS].append(
+        self._make_action('gsutil', '-q', 'cp',
+                          '/google/logs/output',
+                          output_log_path,
+                          flags=['ALWAYS_RUN']))
+    # pylint: disable=no-member
+    request = self._pipeline_service.pipelines().run(body=api_request)
+    operation_name = request.execute()['name']
+    return operation_name
+
+  def _check_and_write_to_output_dir(self, output_dir):
+    # type: (str) -> None
+    real_dir = self._format_dir_path(output_dir)
+    # NOTE(bashir2): We cannot use exists() because for example on GCS, the
+    # directory names are only symbolic and are not physical files.
+    match_results = filesystems.FileSystems.match(['{}*'.format(real_dir)])
+    if match_results and match_results[0].metadata_list:
```

Comment:
I see your point. However, from my DeepVariant experience, users really like just specifying a single output path for all of their runs and don't care about the details. I can bet that users will keep hitting this exception when running the pipeline as they'd forget to assign a unique folder (a common case: user runs program, it fails b/c they don't have the correct VEP cache, they correct it, rerun, the program now fails saying that the log file exists and directory is not empty). Let's discuss to see if there is a way that we can achieve both goals (i.e. provide the flexibility for the user to specify their exact output folder, but also not fail if they dont).

Reply:
BTW, to make sure all comments are replied: As discussed off-PR, while I agree with your point about nuisance of forcing this constraint, I think there is a fundamental difference between this example and DeepVariant. In the DV case I think users almost never care about intermediate files content but for annotations and VEP, those output files are actually valuable independent of BigQuery, specially give the current slow way that we have for running VEP.

Obviously we can/should revisit this and several other decisions we have made about this first version after we get some user feedback.